### PR TITLE
Support for CSR service generated on VaaS

### PR DESCRIPTION
**Pull request by angelmoo**
added support for Support for CSR service generated on VaaS.
added test cases for testing the new enhancement.

### CODE REVIEW

File: pkg/venafi/cloud/cloudUtil.go [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+	if req.Subject.CommonName != "" {
+		if policy.Policy != nil {
+
+			valid, err = isValueMatch([]string{req.Subject.CommonName}, policy.Policy.Domains)
+			if err != nil {
+				return nil, err
+			}
+
+			if !valid {
+				return nil, fmt.Errorf("specified CN %s, doen't match with policy's specified domains %+q", req.Subject.CommonName, policy.Policy.Domains)
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/cloudUtil.go]: fixed on most recent commit

File: pkg/venafi/cloud/cloudUtil.go [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+			if err != nil {
+				return nil, err
+			}
+			if !valid {
+				return nil, fmt.Errorf("specified locality %s, doesn't match with policy's specified localities %+q", req.Subject.Locality[0], policy.Policy.Subject.Localities)
+			}
+		}
+
+		csrAttr.Locality = &req.Subject.Locality[0]
+	} else if policy.Default != nil && policy.Default.Subject != nil && policy.Default.Subject.State != nil {
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/cloudUtil.go]: fixed on most recent commit

File: pkg/venafi/cloud/cloudUtil.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+			if err != nil {
+				return nil, err
+			}
+			if !valid {
+				return nil, fmt.Errorf("specified country %s, doesn't match with policy's specified countries %+q", req.Subject.Country[0], policy.Policy.Subject.Countries)
+			}
+		}
+
+		csrAttr.Country = &req.Subject.Country[0]
+	} else if policy.Default != nil && policy.Default.Subject != nil && policy.Default.Subject.State != nil {
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/cloudUtil.go]: fixed on most recent commit

File: pkg/venafi/cloud/cloud.go [hunk 4]
```diff
@@ -86,6 +87,33 @@ type certificateRequest struct {
...
+}
+
+type KeyStoreRequest struct {
+	ExportFormat                  string `json:"exportFormat,omitempty"`
+	EncryptedPrivateKeyPassphrase string `json:"encryptedPrivateKeyPassphrase"`
+	EncryptedKeystorePassphrase   string `json:"encryptedKeystorePassphrase"`
+	CertificateLabel              string `json:"certificateLabel"`
+}
+
+type DekInfo struct {
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/cloud.go]: fixed on most recent commit

File: pkg/venafi/cloud/connector.go [hunk 5]
```diff
@@ -571,6 +712,20 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+	if req.CertID != "" {
+		dekInfo, err = getDekInfo(c, req.CertID)
+		currentId = req.CertID
+	} else if certificateId != "" {
+		dekInfo, err = getDekInfo(c, certificateId)
+		currentId = certificateId
+	}
+	if err == nil && dekInfo.Key != "" {
+		req.CertID = currentId
+		return retrieveServiceGeneratedCertData(c, req)
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/connector.go]: fixed on most recent commit

File: pkg/venafi/cloud/connector.go [hunk 6]
```diff
@@ -609,6 +764,163 @@ func (c *Connector) RetrieveCertificate(req *certificate.Request) (certificates
...
+			fileBytes, err := ioutil.ReadAll(f)
+			if err != nil {
+				return nil, err
+			}
+
+			certs := strings.Split(strings.TrimSpace(string(fileBytes)), "\n\n")
+
+			for i := 0; i < len(certs); i++ {
+				if i < len(certs)-1 {
+					if chain == "" {
```
[the remark this answers is not in the document]
Comment (angelmoo) [reply pkg/venafi/cloud/connector.go]: fixed on most recent commit

**Review by rvelaVenafi — APPROVED** [review]

**Review by luispresuelVenafi — APPROVED** [review]

**Pull request merged**